### DISCUSSION TITLE
Material UI Select doesn't work as expected after reset

### DISCUSSION THREAD

**Discussion by axisplay-neo** [react-hook-form/react-hook-form#2822] [Q&A]
I would like to use Material UI Select and implement saving on change, and resetting the dirtyFields after reset. 
[codesandbox.io](https://codesandbox.io/s/blazing-night-w8lj1)
However the Material UI Select doesn't work as expected after it is being reset.

The behavior becomes normal if a TextField is added to form, or native select is used. 

Thanks.

**Comment by bluebill1049** [comment]
try with `render` prop

https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r

**Reply by axisplay-neo** [reply]
Thanks. I turn out using more and more controller and render prop for materialUI components.

**Reply by bluebill1049** [reply]
when the external controlled component doesn't expose `ref` then you should use `Controller`

**Reply by kaisermann** [reply]
> when the external controlled component doesn't expose ref then you should use Controller

@bluebill1049 What if the controlled component exposes a ref? 

My components can be controlled/uncontrolled. When they're uncontrolled, they hold an internal state ([comes from this hook](https://github.com/mui-org/material-ui/blob/next/packages/material-ui-utils/src/useControlled.js)). if I use `register` and reset them, their internal state doesn't change. This seems to happen because:

1) no `change` event is triggered when resetting (which is correct), so my `onChange` listen doesn't update the internal value 
2) if `register` was used, the value is reset via something akin to `ref.current.value = defaultValue`

Is the `Controller` wrapper the only workaround?

**Reply by bluebill1049** [reply]
> My components can be controlled/uncontrolled. When they're uncontrolled, they hold an internal state (comes from this hook). if I use register and reset them, their internal state doesn't change. This seems to happen because:

this means your component is not exposed the correct input ref

**Reply by kaisermann** [reply]
Here's a [minimal repro on codesandbox](https://codesandbox.io/s/react-hook-form-custom-input-forked-ididr?file=/src/index.js). Maybe I'm wanting too much or am missing something 🤔 

edit: just hit me that while the component can be uncontrolled, the input is always controlled (internally). duh

**Reply by bluebill1049** [reply]
Don't think you have forwarded the `ref` correctly: https://react-hook-form.com/get-started#Integratinganexistingform

**Reply by kaisermann** [reply]
Oops, my bad, copied from another template. Updated the previous repro on CSB. I think I have convinced myself that it's directly possible to sync that internal value. I Will try to find a way of listening to a form's reset event 🤔
